kong/router/atc_compat.lua

```diff
@@ -0,0 +1,468 @@
+-- +-------------------------+-------------------------------------+
+local function route_priority(r)
+  local match_weight = 0
+
+  if r.methods and #r.methods > 0 then
+    match_weight = match_weight + 1
+  end
+
+  if r.hosts and #r.hosts > 0 then
+    match_weight = match_weight + 1
+  end
+
+  if r.paths and #r.paths > 0 then
+    match_weight = match_weight + 1
+  end
+
+  local headers_count = r.headers and tb_size(r.headers) or 0
+
+  if headers_count > 0 then
+    match_weight = match_weight + 1
+  end
+
+  if headers_count > 255 then
+    ngx_log(ngx_WARN, "too many headers in route ", r.id,
+                      " headers count capped at 255 when sorting")
+    headers_count = 255
+  end
+
+  if r.snis and #r.snis > 0 then
+    match_weight = match_weight + 1
+  end
+
+  local plain_host_only = not not r.hosts
+
+  if r.hosts then
+    for _, h in ipairs(r.hosts) do
+      if h:find("*", nil, true) then
+        plain_host_only = false
+        break
+      end
+    end
+  end
+
+  local max_uri_length = 0
+  local regex_url = false
+
+  if r.paths then
+    for _, p in ipairs(r.paths) do
+      if re_find(p, [[[a-zA-Z0-9\.\-_~/%]*$]], "ajo") then
+        -- plain URI or URI prefix
+        max_uri_length = max(max_uri_length, #p)
+
+      else
+        regex_url = true
+      end
+    end
+  end
+
+  match_weight = lshift(ffi_new("uint64_t", match_weight), 61)
+  headers_count = lshift(ffi_new("uint64_t", headers_count), 52)
+  local regex_priority = lshift(ffi_new("uint64_t", r.regex_priority or 0), 19)
+  local max_length = band(max_uri_length, 0x7FFFF)
+
+  local priority =  bor(match_weight,
+                        plain_host_only and lshift(0x01ULL, 60) or 0,
+                        regex_url and lshift(0x01ULL, 51) or 0,
+                        headers_count,
+                        regex_priority,
+                        max_length)
+
+  return priority
+end
+
+
+function _M.new(routes)
+  if type(routes) ~= "table" then
+    return error("expected arg #1 routes to be a table")
+  end
+
+  local s = schema.new()
+  s:add_field("net.protocol", "String")
+  s:add_field("tls.sni", "String")
+  s:add_field("http.method", "String")
+  s:add_field("http.host", "String")
+  s:add_field("http.path", "String")
+  s:add_field("http.raw_path", "String")
+  s:add_field("http.headers.*", "String")
+
+  local inst = router.new(s)
+
+  local router = setmetatable({
+    schema = s,
+    router = inst,
+    routes = {},
+    services = {},
+    fields = nil,
+  }, _MT)
+
+  for _, r in ipairs(routes) do
+    router.routes[r.route.id] = r.route
+    router.services[r.route.id] = r.service
+
+    get_atc(r.route)
+    assert(inst:add_matcher(route_priority(r.route), r.route.id, get_atc(r.route)))
+
+    router.fields = inst:get_fields()
+  end
+
+  return router
+end
+
+
+function _M:select(req_method, req_uri, req_host, req_scheme,
+                   _src_ip, _src_port,
```

Comment:
wrong params name `_src_ip` `_src_port`

Reply:
It is necessary in order to be compatible with existing router. These are unused variables thus prefixed with `_`.

Reply:
how about add a comment to describe this?